src/cms/contentful.ts

```diff
@@ -14,80 +15,128 @@ import {
+   * const entry = await client.getEntry<MyEntry>({ entryId: '123', locale: 'en-US' });
+   * console.log(entry); // { data: [{ title: 'My title', description: 'My description' }] }
+   */
+  public async getEntry<T = Record<string, unknown>>({
     entryId,
     locale,
     nestedLevels = 10,
-  }: ContentfulGetEntryParams): Promise<HarmonizedOutput> {
+  }: ContentfulGetEntryParams): Promise<HarmonizedOutput<T>> {
     const query = {
       locale,
-      include: nestedLevels, // Include up to 10 levels of linked entries
+      include: nestedLevels,
     };
 
-    return await this.getEntryHarmonized(
-      this.getEntryHandler.bind(this, entryId, query),
-      this.parserHandler.bind(this),
+    return await this.getEntryHarmonized<T>(
+      <() => Promise<T>>this.getEntryHandler?.bind(this, entryId, query),
+      <(data: T) => HarmonizedOutput<T>>this.parserHandler.bind(this),
     );
   }
 
-  private async getEntryHandler(
+  protected clientInstance: ContentfulClientApi<undefined>;
+
+  protected getClientInstance(): ContentfulClientApi<undefined> {
+    return this.clientInstance;
+  }
+
+  private async getEntryHandler<T = Record<string, unknown>>(
     entryId: string,
     query?: EntryQueries<undefined>,
-  ): Promise<Entry<EntrySkeletonType, undefined, string>> {
-    return this.getClientInstance().getEntry(entryId, query);
+  ): Promise<ContentfulEntry<T>> {
+    return this.getClientInstance().getEntry<ContentfulEntrySkeleton<T>>(
+      entryId,
+      query,
+    );
   }
 
-  private parserHandler(
-    payload: Entry<EntrySkeletonType, undefined, string>,
-  ): HarmonizedOutput {
-    const output = this.mapper(payload);
+  private parserHandler<T = Record<string, unknown>>(
+    payload: ContentfulEntry<T>,
+  ): HarmonizedOutput<string | ContentfulEntry<T>> {
+    const output = this.mapper<T>(payload);
     return { data: output };
   }
 
-  private mapper(item: any): any {
-    switch (item?.sys?.type) {
+  private mapper<T = Record<string, unknown>>(
+    item: ContentfulEntry<T>,
+  ): string | ContentfulEntry<T> {
+    switch (item?.sys?.type + '') {
       case ContentfulResourceType.ENTRY:
         return Object.entries(item.fields).reduce((acc, [key, value]) => {
           if (Array.isArray(value)) {
             return {
               ...acc,
-              [key]: value.map((valueItem) => this.mapper(valueItem)),
+              [key]: value.map((valueItem) =>
+                this.mapper<T>(valueItem || Object.create(null)),
+              ),
             };
           }
-          return { ...acc, [key]: this.mapper(value) };
-        }, {});
+          return {
+            ...acc,
+            [key]: this.mapper<T>(value || Object.create(null)),
+          };
+        }, Object.create(null));
 
       case ContentfulResourceType.ASSET:
-        return `https:${item.fields?.file?.url}`;
+        return `https:${Object(item.fields?.file)?.url}`;
```

Comment:
If you check the type definition of `item.fields.file` from the Contentful lib the result will be:
`string | number | boolean | EntryFields.Location | Document | JsonObject | JsonArray | UnresolvedLink<"Asset"> | ... 5 more ... | null` which hints that `url` is not available in all cases. To get around without further abstraction I simply cast the base prop to an untyped object. In this case, this is totally safe, since all values can be casted to an object. If we swap the idea and omit the `Object` construction, we might run into runtime errors like `Property url does not exists on null` or something. Hope this makes sense :)